### DISCUSSION TITLE
ComfyUI Master Manager (PowerShell)

### DISCUSSION THREAD

**Discussion by Rogala** [Comfy-Org/ComfyUI#12675] [Show and tell]
Reliable Version & Environment Control

Prerequisites & Compatibility
This tool is specifically designed for:

Git-based Installations: You must have a local version of ComfyUI cloned via git clone.

Virtual Environments (Venv): The script expects a standard Python virtual environment (located in the ./venv/ folder relative to the script).

Standard Directory Structure: It targets a setup where ComfyUI and the venv folder reside in the same root directory.

[!IMPORTANT]
This manager is not compatible with "Portable/Embedded" ComfyUI distributions that use internal python runners.
[update.zip](https://github.com/user-attachments/files/25605102/update.zip)
<img width="192" height="170" alt="file" src="https://github.com/user-attachments/assets/59a0d16e-ff18-4e50-8093-5192e7dc3b22" />

Overview
This manager is a robust tool for users who need to switch between different ComfyUI releases or hardware backends (Nvidia, Intel, CPU) without breaking their Python environment. It focuses on stability and clean installation, preventing common "dependency hell" issues.

How to Run
Launcher: Run via update.bat.

System: Works on Windows PowerShell 5.1+.

Automation: It automatically bypasses execution policies to ensure a smooth start.

Key Features
1. Smart Environment Diagnostics
Quickly check your current setup, including ComfyUI version, Python details, and full Torch/CUDA status. It even monitors your Pip cache to help manage disk space.

2. Conflict-Free Version Switching
Switch between ComfyUI releases with confidence. The script doesn't just change the code; it performs an automated compatibility check and patches critical system libraries to prevent startup warnings and network errors common in older versions.

3. Unified Hardware Stack Management
Switch between NVIDIA (CUDA), Intel (XPU), or CPU modes.

Simplified Selection: The menu shows clean, easy-to-read version numbers.

Deep Sync: It forces all core components to update simultaneously, ensuring they are perfectly matched for your chosen hardware.

Future-Proofing: The script is designed to be easily updated. If new CUDA versions are released, the developer can simply update the $cu array (e.g., @("126", "128", "130")) with values from the official [PyTorch Get Started](https://pytorch.org/get-started/locally/) page to support the latest drivers.

Clean Reinstalls: Automatically removes conflicting leftovers when switching between different GPU types.

4. Integrated Release Intelligence
Before switching versions, you can view the official Release Notes directly in the console to see what’s new or what might have changed.

5. Maintenance & Logs
Includes automated cleanup of temporary lock files that often cause "Permission Denied" errors during updates. All actions are logged for easy troubleshooting.

Summary of Benefits
No more "RequestsDependencyWarning" or broken network nodes.

No more mismatched CUDA versions that prevent ComfyUI from seeing your GPU.

One-click launch with a clean, vertical menu interface.

<details>
  <summary>File update.bat</summary>

@echo off
start powershell -NoProfile -ExecutionPolicy Bypass -NoExit -File "%~dp0update_comfy.ps1"

</details>

<details>
  <summary>update_comfy.ps1</summary>

# =========================================================
# ComfyUI Master Manager v6.6 (Final Stable Edition)
# Interface: Vertical English | Support: CUDA 126, 128, 130
# Features: Nuclear Dep-Patch, Clean Versioning, Auto-Cache
# =========================================================

# ------------------ MODULE 1: Initialization ------------------
$ScriptRoot = Split-Path -Parent $MyInvocation.MyCommand.Definition
$CacheDir   = Join-Path $ScriptRoot ".cache"
if (!(Test-Path $CacheDir)) { New-Item $CacheDir -ItemType Directory | Out-Null }

$LogFile = Join-Path $CacheDir "history.log"
$rnFile  = Join-Path $CacheDir "release_notes.log"
$VenvPython = Join-Path $ScriptRoot "venv\Scripts\python.exe"
$ComfyDir   = Join-Path $ScriptRoot "ComfyUI"
$SitePackages = Join-Path $ScriptRoot "venv\Lib\site-packages"

$OutputEncoding = [System.Text.UTF8Encoding]::new()

function Write-Log([string]$msg) {
    "$(Get-Date -Format 'yyyy-MM-dd HH:mm:ss') | $msg" | Add-Content $LogFile
}

function Cleanup-Venv {
    if (Test-Path $SitePackages) {
        Get-ChildItem $SitePackages -Filter "~*" -ErrorAction SilentlyContinue | Remove-Item -Recurse -Force -ErrorAction SilentlyContinue
    }
}

# ------------------ MODULE 2: UI Engine ------------------
function Read-Choice([string]$Prompt, [int[]]$Allowed) {
    while ($true) {
        $input = Read-Host $Prompt
        if ($input -match '^\d+$' -and $Allowed -contains [int]$input) { return [int]$input }
        Write-Host "[!] Error: Invalid selection." -ForegroundColor Red
    }
}

# ------------------ MODULE 3: Show Environment Info ------------------
function Show-Info {
    Cleanup-Venv
    Write-Host "`n=== SYSTEM ENVIRONMENT INFO ===" -ForegroundColor White -BackgroundColor DarkBlue
    
    $tag = if (Test-Path $ComfyDir) { 
        Push-Location $ComfyDir; $t = git describe --tags --abbrev=0 2>$null; Pop-Location; $t 
    } else { "Not found" }
    
    $pyV = if (Test-Path $VenvPython) { (& $VenvPython --version 2>&1).ToString().Trim() } else { "N/A" }
    
    $torchCode = "import torch; print(f'VER:{torch.__version__}|CUDA:{torch.version.cuda}|AVAIL:{torch.cuda.is_available()}')"
    $tInfo = if (Test-Path $VenvPython) { (& $VenvPython -c $torchCode 2>$null) } else { "N/A" }
    
    Write-Host " ComfyUI Version: " -NoNewline; Write-Host $tag -ForegroundColor Cyan
    Write-Host " Python Version:  " -NoNewline; Write-Host $pyV -ForegroundColor Cyan
    
    if ($tInfo -match "VER:(.*?)\|CUDA:(.*?)\|AVAIL:(.*)") {
        Write-Host " Torch Version:   " -NoNewline; Write-Host $Matches[1] -ForegroundColor Cyan
        Write-Host " CUDA Version:    " -NoNewline; Write-Host $Matches[2] -ForegroundColor Cyan
        $color = if ($Matches[3] -eq "True") { "Green" } else { "Red" }
        Write-Host " CUDA Supported:  " -NoNewline; Write-Host $Matches[3] -ForegroundColor $color
    }

    Write-Host "`n--- Pip Cache Status (Total) ---" -ForegroundColor Gray
    if (Test-Path $VenvPython) {
        & $VenvPython -m pip cache info | ForEach-Object {
            if ($_ -match "Size:") { Write-Host " Pip Cache $_" -ForegroundColor Yellow }
        }
    }
    Write-Host "==============================="
    Read-Host "`nPress Enter to return to menu"
}

# ------------------ MODULE 4: ComfyUI Management ------------------
function Change-ComfyUI {
    Cleanup-Venv
    if (!(Test-Path $ComfyDir)) { Write-Host "[ERROR] ComfyUI directory not found!" -ForegroundColor Red; return }
    
    Push-Location $ComfyDir
    Write-Host "`n[GIT] Fetching tags..." -ForegroundColor Gray
    git fetch --tags --all --quiet
    
    $tags = git tag -l --sort=-v:refname | Select-Object -First 5
    $current = git describe --tags --abbrev=0 2>$null

    Write-Host "`nShow release notes for these versions?"
    Write-Host " [1] Yes`n [0] No"
    if ((Read-Choice "Select" @(0,1)) -eq 1) {
        foreach ($tag in $tags) {
            try {
                $res = Invoke-RestMethod "https://api.github.com/repos/comfyanonymous/ComfyUI/releases/tags/$tag" -ErrorAction SilentlyContinue
                if ($res) { 
                    $cleanBody = ($res.body -split '## New Contributors' | Select-Object -First 1).Trim()
                    Write-Host "`n--- $tag ---" -ForegroundColor Cyan
                    Write-Host $cleanBody -ForegroundColor Gray
                    "=== $tag ===`n$cleanBody`n" | Add-Content $rnFile
                }
            } catch { Write-Host "Failed to fetch notes for $tag" -ForegroundColor DarkGray }
        }
    }

    Write-Host "`nAvailable ComfyUI Versions:"
    for ($i=0; $i -lt $tags.Count; $i++) {
        $mark = if ($tags[$i] -eq $current) { "(current)" } else { "" }
        Write-Host " [$i] $($tags[$i]) $mark"
    }
    Write-Host " [9] Back"

    $choice = Read-Choice "Select version" (0..($tags.Count-1) + 9)
    if ($choice -eq 9) { Pop-Location; return }

    $target = $tags[$choice]
    Write-Host "`n[PROCESS] Switching to $target..." -ForegroundColor Yellow
    git checkout $target --force --quiet
    
    Write-Host "[PROCESS] Syncing dependencies (no-deps)..." -ForegroundColor Yellow
    & $VenvPython -m pip install -r requirements.txt --upgrade --no-deps --prefer-binary
    
    # PATCH: NUCLEAR FIX for RequestsDependencyWarning. Pins all related libs to legacy stable versions.
    Write-Host "[PROCESS] Patching requests compatibility stack..." -ForegroundColor Gray
    & $VenvPython -m pip install "requests==2.31.0" "urllib3==1.26.18" "chardet==5.2.0" "charset-normalizer==3.3.2" --prefer-binary
    
    Write-Host "`n[SUCCESS] ComfyUI version changed to $target" -ForegroundColor Green
    Write-Log "ComfyUI switched: $current -> $target"
    Read-Host "Press Enter to continue"
    Pop-Location
}

# ------------------ MODULE 5: Torch Management ------------------
function Change-Torch {
    Cleanup-Venv
    if (!(Test-Path $VenvPython)) { Write-Host "[ERROR] venv not found!" -ForegroundColor Red; return }

    $currentTorch = (& $VenvPython -c "import torch; print(torch.__version__)" 2>$null)

    Write-Host "`n--- Torch Stack Configuration ---" -ForegroundColor Cyan
    Write-Host " [0] CPU`n [1] NVIDIA CUDA`n [2] Intel XPU`n [9] Back"
    $dev = Read-Choice "Select device" @(0,1,2,9)
    if ($dev -eq 9) { return }

    $url = switch ($dev) {
        1 { 
			# To support future CUDA releases, update this array with values from https://pytorch.org/get-started/locally/
            $cu = @("126", "128", "130")
            Write-Host "`nSelect CUDA version:"
            for($i=0; $i -lt $cu.Count; $i++) { Write-Host " [$i] cu$($cu[$i])" }
            "https://download.pytorch.org/whl/cu$($cu[(Read-Choice 'Choice' @(0,1,2))])"
        }
        2 { "https://download.pytorch.org/whl/xpu" }
        default { "https://download.pytorch.org/whl/cpu" }
    }

    Write-Host "`nFetching available versions..." -ForegroundColor Gray
    $raw = cmd /c "`"$VenvPython`" -m pip install torch==0.0.0 --index-url $url 2>&1"
    if ($raw -join " " -match "from versions: (.*)\)") {
        $versions = $Matches[1].Split(',') | ForEach-Object { $_.Trim() } | 
                    Where-Object { $_ -match '^\d+\.\d+\.\d+' } | 
                    Select-Object -Unique | Sort-Object {[version]($_ -replace '\+.*','')} -Descending | 
                    Select-Object -First 5
        
        Write-Host "`nSelect Torch version:"
        for ($i=0; $i -lt $versions.Count; $i++) { 
            $cleanName = $versions[$i] -replace '\+.*',''
            $mark = if ($versions[$i] -eq $currentTorch) { "(current)" } else { "" }
            Write-Host " [$i] $cleanName $mark"
        }
        Write-Host " [9] Back"

        $vIndex = Read-Choice "Choice" (0..($versions.Count-1) + 9)
        if ($vIndex -eq 9) { return }
        $target = $versions[$vIndex]
        
        Write-Host "`n[PROCESS] Installing Torch Stack ($target)..." -ForegroundColor Yellow
        & $VenvPython -m pip install "torch==$target" torchvision torchaudio --index-url $url --upgrade --prefer-binary --force-reinstall
        
        Write-Host "[PROCESS] Finalizing requirements..." -ForegroundColor Yellow
        if (Test-Path (Join-Path $ComfyDir "requirements.txt")) {
            & $VenvPython -m pip install -r (Join-Path $ComfyDir "requirements.txt") --no-deps --prefer-binary
        }
        
        Write-Log "Torch switched: $currentTorch -> $target"
        Write-Host "`n[SUCCESS] Torch stack updated to $target" -ForegroundColor Green
        Read-Host "Press Enter to continue"
    }
}

# ------------------ MAIN LOOP ------------------
while ($true) {
    Clear-Host
    Write-Host "`n=== COMFYUI MANAGER v6.6 ===" -ForegroundColor White -BackgroundColor DarkMagenta
    Write-Host " [1] Change ComfyUI Version (Tags/Notes)"
    Write-Host " [2] Change Torch Stack (CPU/Nvidia/Intel)"
    Write-Host " [3] Show Environment Info"
    Write-Host " [9] Exit"

    switch (Read-Choice "Select option" @(1,2,3,9)) {
        1 { Change-ComfyUI }
        2 { Change-Torch }
        3 { Show-Info }
        9 { exit }
    }
}

</details>
